### PR TITLE
HIVE-23750: Rewrite plan to join back tables: support function calls in project

### PR DESCRIPTION
**Pull request by kasakrisz**
Testing done:
```
mvn test -Dtest.output.overwrite -DskipSparkTests -Dtest=TestMiniLlapLocalCliDriver -Dqfile=cardinality_preserving_join_opt.q -pl itests/qtest -Pitests
mvn test -Dtest.output.overwrite -DskipSparkTests -Dtest=TestTezPerfConstraintsCliDriver -Dqfile=cbo_query4.q,cbo_query11.q,cbo_query74.q -pl itests/qtest -Pitests
```

### CODE REVIEW

File: ql/src/test/results/clientpositive/perf/tez/constraints/cbo_query84.q.out [hunk 1]
```diff
@@ -56,24 +56,24 @@ CBO PLAN:
 HiveProject(customer_id=[$0], customername=[$1])
   HiveSortLimit(sort0=[$2], dir0=[ASC], fetch=[100])
     HiveProject(customer_id=[$2], customername=[$6], c_customer_id=[$2])
-      HiveJoin(condition=[=($8, $4)], joinType=[inner], algorithm=[none], cost=[not available])
-        HiveJoin(condition=[=($1, $3)], joinType=[inner], algorithm=[none], cost=[not available])
-          HiveJoin(condition=[=($0, $1)], joinType=[inner], algorithm=[none], cost=[not available])
+      HiveJoin(condition=[=($8, $4)], joinType=[inner], algorithm=[CommonJoin], cost=[not available])
```
Comment (jcamachor) [inline ql/src/test/results/clientpositive/perf/tez/constraints/cbo_query84.q.out:59]: This is not nice, it seem a side effect of using the new cost model. I am wondering why we did not hit it before? Are we changing the metadata provider correctly?

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveCardinalityPreservingJoinOptimization.java [hunk 2]
```diff
@@ -161,127 +174,118 @@ public RelNode trim(RelBuilder relBuilder, RelNode root) {
...
       List<RexInputRef> projectExpressions, RelNode projectInput) {
     RelMetadataQuery relMetadataQuery = RelMetadataQuery.instance();
-    Map<RexTableInputRef.RelTableRef, ProjectedFieldsBuilder> fieldMappingBuilders = new HashMap<>();
+    Map<RexTableInputRef.RelTableRef, JoinedBackFieldsBuilder> fieldMappingBuilders = new HashMap<>();
     List<RexTableInputRef.RelTableRef> tablesOrdered = new ArrayList<>(); // use this list to keep the order of tables
     for (RexInputRef expr : projectExpressions) {
       Set<RexNode> expressionLineage = relMetadataQuery.getExpressionLineage(projectInput, expr);
       if (expressionLineage == null || expressionLineage.size() != 1) {
-        LOG.debug("Lineage can not be determined of expression: " + expr);
+        LOG.debug("Lineage of expression can not be determined: " + expr);
```
[the remark this answers is not in the document]
Comment (kasakrisz) [reply ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveCardinalityPreservingJoinOptimization.java]: fixed

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveCardinalityPreservingJoinOptimization.java [hunk 3]
```diff
@@ -161,127 +174,118 @@ public RelNode trim(RelBuilder relBuilder, RelNode root) {
...
-      RexNode[] newProjects = new RexNode[rootFieldList.size()];
-      String[] newColumnNames = new String[rootFieldList.size()];
-      projectsFromOriginalPlan(rexBuilder, newInput.getRowType().getFieldCount(), newInput, newInputMapping,
-          newProjects, newColumnNames);
+      Map<RexTableInputRef, Integer> tableInputRefMapping = new HashMap<>();
 
-      // 5. Join back tables to the top of original plan
       for (TableToJoinBack tableToJoinBack : tableToJoinBackList) {
-        LOG.debug("Joining back table " + tableToJoinBack.projectedFields.relOptHiveTable.getName());
+        LOG.debug("Joining back table " + tableToJoinBack.joinedBackFields.relOptHiveTable.getName());
```
[the remark this answers is not in the document]
Comment (kasakrisz) [reply ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveCardinalityPreservingJoinOptimization.java]: fixed

File: ql/src/test/results/clientpositive/perf/tez/constraints/cbo_query84.q.out (hunk 1 above)
Comment (kasakrisz) [reply ql/src/test/results/clientpositive/perf/tez/constraints/cbo_query84.q.out]: Added logic to reset the `joinAlgorithm` back to `none` in all `HiveJoin` nodes after calculating the cost and switching back the MetadataProvider

**Pull request merged**